From f41bc530a8b1a66e811e5c3b4da3df72ecba0e1d Mon Sep 17 00:00:00 2001
From: Jeremy Harris <jgh146exb@wizmail.org>
Date: Wed, 26 Jun 2019 11:17:52 +0100
Subject: [PATCH 28/34] Fix DSN Final-Recipient: field

(cherry picked from commits 436bda2ac0c4 and 98d4eb7a84)

(cherry picked from commit 6b88f51ac13b4fa834796ce12d12c55c95eacc4a)
(cherry picked from commit ce4d8eca9d3940bb439cdb74a250090fee5538d4)
---
 doc/ChangeLog     |  5 +++++
 src/deliver.c         | 39 +++++++++++++++++++++++++--------------
 test/mail/0015.CALLER     |  4 ++--
 test/mail/0037.CALLER     |  4 ++--
 test/mail/0048.CALLER     | 10 +++++-----
 test/mail/0098.CALLER     |  6 +++---
 test/mail/0103.CALLER     |  2 +-
 test/mail/0136.forwarder  |  4 ++--
 test/mail/0164.CALLER     |  4 ++--
 test/mail/0165.CALLER     |  4 ++--
 test/mail/0224.CALLER     | 10 +++++-----
 test/mail/0226.CALLER     |  6 +++---
 test/mail/0237.CALLER     |  6 +++---
 test/mail/0383.CALLER     |  6 +++---
 test/mail/0428.CALLER     |  4 ++--
 test/mail/0508.postmaster |  2 +-
 16 files changed, 66 insertions(+), 50 deletions(-)

diff --git a/doc/ChangeLog b/doc/ChangeLog
index 67dfafa7a..3b36c87b9 100644
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -82,6 +82,11 @@ JH/27 Bug 2404: Use the main-section configuration option "dsn_from" for
 
 PP/01 Unbreak heimdal_gssapi, broken in 4.92.
 
+JH/29 Fix DSN Final-Recipient: field.  Previously it was the post-routing
+      delivery address, which leaked information of the results of local
+      forwarding.  Change to the original envelope recipient address, per
+      standards.
+
 
 Exim version 4.92
 -----------------
diff --git a/src/deliver.c b/src/deliver.c
index 4720f596a..53562dd5c 100644
--- a/src/deliver.c
+++ b/src/deliver.c
@@ -5507,6 +5507,25 @@ while ((addr = *anchor))
 
 
 
+/************************************************/
+
+static void
+print_dsn_addr_action(FILE * f, address_item * addr,
+  uschar * action, uschar * status)
+{
+address_item * pa;
+
+if (addr->dsn_orcpt)
+  fprintf(f,"Original-Recipient: %s\n", addr->dsn_orcpt);
+
+for (pa = addr; pa->parent; ) pa = pa->parent;
+fprintf(f, "Action: %s\n"
+    "Final-Recipient: rfc822;%s\n"
+    "Status: %s\n",
+  action, pa->address, status);
+}
+
+
 /*************************************************
 *              Deliver one message               *
 *************************************************/
@@ -7410,10 +7429,7 @@ if (addr_senddsn)
       if (addr_dsntmp->dsn_orcpt)
         fprintf(f,"Original-Recipient: %s\n", addr_dsntmp->dsn_orcpt);
 
-      fprintf(f, "Action: delivered\n"
-	  "Final-Recipient: rfc822;%s\n"
-	  "Status: 2.0.0\n",
-	addr_dsntmp->address);
+      print_dsn_addr_action(f, addr_dsntmp, US"delivered", US"2.0.0");
 
       if (addr_dsntmp->host_used && addr_dsntmp->host_used->name)
         fprintf(f, "Remote-MTA: dns; %s\nDiagnostic-Code: smtp; 250 Ok\n\n",
@@ -7798,10 +7814,9 @@ wording. */
       for (addr = handled_addr; addr; addr = addr->next)
         {
 	host_item * hu;
-        fprintf(fp, "Action: failed\n"
-	    "Final-Recipient: rfc822;%s\n"
-	    "Status: 5.0.0\n",
-	    addr->address);
+
+	print_dsn_addr_action(fp, addr, US"failed", US"5.0.0");
+
         if ((hu = addr->host_used) && hu->name)
 	  {
 	  fprintf(fp, "Remote-MTA: dns; %s\n", hu->name);
@@ -8343,13 +8358,9 @@ else if (addr_defer != (address_item *)(+1))
 
         for ( ; addr_dsndefer; addr_dsndefer = addr_dsndefer->next)
           {
-          if (addr_dsndefer->dsn_orcpt)
-            fprintf(f, "Original-Recipient: %s\n", addr_dsndefer->dsn_orcpt);
 
-          fprintf(f, "Action: delayed\n"
-	      "Final-Recipient: rfc822;%s\n"
-	      "Status: 4.0.0\n",
-	    addr_dsndefer->address);
+	  print_dsn_addr_action(f, addr_dsndefer, US"delayed", US"4.0.0");
+
           if (addr_dsndefer->host_used && addr_dsndefer->host_used->name)
             {
             fprintf(f, "Remote-MTA: dns; %s\n",
-- 
2.23.0.rc1

